tests/device_selector/device_selector_aspect.cpp

```diff
@@ -0,0 +1,469 @@
+#ifndef SYCL_CTS_COMPILING_WITH_DPCPP
+  test_selector(sycl::aspect_selector(accept_list), accept_list);
+  if (!deny_list.empty()) {
+    test_selector(sycl::aspect_selector(accept_list, deny_list), accept_list,
+                  deny_list);
+  }
+  test_selector(sycl::aspect_selector(Aspects...), accept_list);
+  test_selector(sycl::aspect_selector<Aspects...>(), accept_list);
+#endif  // SYCL_CTS_COMPILING_WITH_DPCPP
+};
+
+/** Functor that checks a selector with one aspect, no denied aspects */
+template <typename AspectT>
+class check_for_single_aspect {
+  static constexpr sycl::aspect aspect = AspectT::value;
+
+ public:
+  void operator()(const std::string& aspect_name) {
+    INFO("for aspect " << aspect_name);
+    check_aspect_selector<aspect>({});
+  }
+};
+
+/** Functor that checks a selector with two aspects, no denied aspects */
+template <typename Aspect1T, typename Aspect2T>
+class check_for_two_aspects {
+  static constexpr sycl::aspect aspect1 = Aspect1T::value;
+  static constexpr sycl::aspect aspect2 = Aspect2T::value;
+
+ public:
+  void operator()(const std::string& aspect1_name,
+                  const std::string& aspect2_name) {
+    INFO("for aspects " << aspect1_name << " and " << aspect2_name);
+    check_aspect_selector<aspect1, aspect2>({});
+  }
+};
+
+/**
+ * Returns a randomly-sized list of random denied aspects that are not part of
+ * the requested aspects \p selected_aspect_list. */
+template <typename Rng>
+std::vector<sycl::aspect> generate_denied_list(
+    const std::vector<sycl::aspect>& aspect_list,
+    const std::vector<sycl::aspect>& selected_aspect_list, Rng& rng) {
+  std::vector<sycl::aspect> denied_list;
+  for (unsigned int i = 0; i < aspect_list.size(); i++) {
+    bool is_selected =
+        std::find(selected_aspect_list.begin(), selected_aspect_list.end(),
+                  aspect_list[i]) != selected_aspect_list.end();
+    if (is_selected) {
+      continue;
+    }
+    unsigned int p = rng() % aspect_list.size();
+    if (p == 0) {
+      denied_list.push_back(aspect_list[i]);
+    }
+  }
+  return denied_list;
+}
+
+/**
+ * Functor that checks a selector with multiple aspects and optionally
+ * a list of denied aspects. */
+template <typename... AspectsT>
+class check_for_multiple_aspects {
+  static constexpr std::size_t aspect_count = sizeof...(AspectsT);
+
+ public:
+  void operator()(const std::vector<sycl::aspect>& deny_list,
+                  const std::string accept_aspect_names[aspect_count]) {
+    std::ostringstream os;
+    os << "for aspects:\n";
+    for (std::size_t i = 0; i < aspect_count; i++) {
+      os << accept_aspect_names[i] << "\n";
+    }
+    os << "for denied aspects:\n";
+    for (const sycl::aspect& aspect : deny_list) {
+      os << Catch::StringMaker<sycl::aspect>::convert(aspect) << "\n";
+    }
+    INFO(os.str());
+
+    check_aspect_selector<AspectsT::value...>(deny_list);
+  }
+};
+
+template <typename... AspectsT>
+struct helper_subset {
```

Comment:
All the following should be modernized by using `std::make_index_sequence` https://en.cppreference.com/w/cpp/utility/integer_sequence instead of using old-style metaprogramming.
Probably adding to `util/variadic.h` a kind of constexpr for loop taking higher level function would allow some code factorizing.

Reply:
Actually there is a similar change which has been done in https://github.com/KhronosGroup/SYCL-CTS/pull/442/commits/debb378eb024fddb540188b5d655dfb4afa8102f
Perhaps this could be also simplified once the mechanics is moved to `variadic.h`.

Reply:
I have replaced the `helper_subset` with a modern metaprogramming approach. The `helper_random` cannot be replaced in the same way, due to the recursive nature of the random number generator required for selecting random aspects. I do not think this recursion can be achieved with parameter pack expansion and/or fold expressions.

I don't think it is valuable to extract some common functionality to a header file. As far as I can see, the only thing that can be done here, is extracting this pattern:
```c++
template<std::size_t Indices...>
void bar(std::index_sequence<Indices...>)
{
  // some use of Indices
}

void foo()
{
  bar(std::make_index_sequence<N>{});
}
```
However, I think this pattern is so simple that extracting it would make the code harder to read.

Please let me know if you disagree with any of these statements. It is evident (also from the linked comment) that I am catching up to metaprogramming beyond C++11, so I'd love to hear your feedback.

Reply:
The random generator is recursive because you chose so. For example you could just use the index as a seed. This is what is done with cryptographic random generator.
Perhaps we can wait for C++20 and 23 before refactoring all this code since it will be even simpler.

Reply:
> The random generator is recursive because you chose so. For example you could just use the index as a seed. This is what is done with cryptographic random generator.

That is absolutely true. I had considered something like Philox or Threefry but I decided against it as that would be overkill. I have now updated the generation of random subsets for modern metaprogramming by creating a unique seed for each element that is generated. This is far from perfect randomness but I really do not think that is required in this situation.